ChiModules/KEigenvalueSolver/k_eigenvalue_execute.cc

```diff
@@ -0,0 +1,23 @@
+#include "k_eigenvalue_solver.h"
+
+#include <chi_mpi.h>
+#include <chi_log.h>
+
+extern ChiMPI& chi_mpi;
+extern ChiLog& chi_log;
+
+#include <iomanip>
+
+
+using namespace LinearBoltzmann;
+
+//###################################################################
+void KEigenvalue::Solver::ExecuteKSolver()
+{
+  MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
Again, I don't think this barrier or the one below are needed.

Reply:
Mmmm interesting. Friendly question: Are the barrier concerns hard objections or just mild concerns? I am asking because we are still going strong in developing the code and when I see barriers that would obviously cause performance concerns then I would also advise them to be "off with his head". Otherwise, why not leave them? In this case especially, I would actually follow some PETSc recommendations and argue that execute is a HUGE step. Wrap it like this:

```c++
void KEigenvalue::Solver::ExecuteKSolver()
{
  chi_log.Log() << "KEigenvalue::Solver execution started\n";
  MPI_Barrier(MPI_COMM_WORLD);

  PowerIteration();

  chi_log.Log() << "KEigenvalue::Solver execution completed\n";
  MPI_Barrier(MPI_COMM_WORLD);
}
```

What I've seen happening is that `execute` completes and then one of the processors are slightly ahead and encounters a bug in field function exporting routines. Then the developer spends endless amounts of time just to discover that the solver is not to blame.

Reply:
In general, I think it's poor practice to add barriers without a very compelling reason. Lawrence and I didn't see eye-to-eye on a lot, but, on this point, I think we would be in complete agreement. In this case, I don't see the need for the initial barrier (what operation are we enforcing synchronization for?), and the second barrier is essentially a duplicate barrier (coming almost immediately after the implicit barrier in the all-reduce in PowerIteration). I do understand the need to add barriers for debugging (or as you say training-wheel purposes). In those cases, the reason should be well documented so that we can remove them later.